tensorflow/core/common_runtime/pluggable_device/pluggable_device.cc

```diff
@@ -0,0 +1,425 @@
+    Allocator* cpu_allocator, bool sync_every_op)
+    : LocalDevice(options, Device::BuildDeviceAttributes(
+                               name, device_type.c_str(), memory_limit,
+                               locality, physical_device_desc)),
+      device_allocator_(device_allocator),
+      cpu_allocator_(cpu_allocator),
+      tf_device_id_(tf_device_id),
+      platform_name_(platform_name),
+      sync_every_op_(sync_every_op) {
+  if (options.config.has_gpu_options()) {
+    force_gpu_compatible_ = options.config.gpu_options().force_gpu_compatible();
+  }
+  PluggableDeviceProcessState::singleton(device_type, platform_name)
+      ->EnablePluggableDevice();
+}
+
+PluggableDevice::~PluggableDevice() {
+  delete pluggable_device_info_;
+  device_context_->Unref();
+}
+
+Status PluggableDevice::Init(const SessionOptions& options) {
+  se::Platform* platform = PluggableDeviceMachineManager(platform_name_);
+  auto executor_status = DeviceIdUtil::ExecutorForTfDeviceId(
+      DeviceType(device_type()), platform, tf_device_id_);
+  if (!executor_status.status().ok()) {
+    return errors::Internal("Failed to get StreamExecutor for device",
+                            tf_device_id_.value());
+  }
+  executor_ = executor_status.ValueOrDie();
+
+  em_ = EventMgrFactory::Singleton()->GetEventMgr(executor_,
+                                                  options.config.gpu_options());
+
+  stream_ = StreamGroupFactory::Global().GetOrCreate(
+      device_type(), tf_device_id_, 0, executor_, options.config.gpu_options());
+  device_context_ = new PluggableDeviceContext(
+      0, stream_->compute, stream_->host_to_device, stream_->device_to_host,
+      stream_->device_to_device);
+  pluggable_device_info_ = new GpuDeviceInfo;
+  pluggable_device_info_->stream = stream_->compute;
+  pluggable_device_info_->default_context = device_context_;
+  pluggable_device_info_->event_mgr = em_;
+  PlatformDeviceId platform_device_id;
+  TF_RETURN_IF_ERROR(DeviceIdManager::TfToPlatformDeviceId(
+      DeviceType(device_type()), tf_device_id_, &platform_device_id));
+  pluggable_device_info_->gpu_id = platform_device_id.value();
+  set_tensorflow_gpu_device_info(pluggable_device_info_);
+
+  // Whether and how the PluggableDevice uses its own threadpool.
+  // This option is experimental. Once we confirm the best setting, we
+  // may change the default behavior and completely remove this flag.
+  // Default values might change in future releases.
+  // Possible values:
+  //   * global: PluggableDevice uses threads shared with CPU in the main
+  //       compute thread-pool. This is currently the default.
+  //   * gpu_private: PluggableDevice uses threads dedicated to this device.
+  //   * gpu_shared: All PluggableDevice s share a dedicated thread pool.
+  string gpu_thread_mode;
+  TF_RETURN_IF_ERROR(
+      ReadStringFromEnvVar("TF_GPU_THREAD_MODE", "global", &gpu_thread_mode));
```

Comment:
Ideally this would be a configuration returned by a pluggable device callback (which can query whatever env var it needs).

For now I'm OK with `TF_GPU_THREAD_MODE` / `TF_GPU_THREAD_COUNT`

Reply:
That sounds great. :)

@jzhoulon Could you please add a TODO to replace these with configs from PluggableDevice? 
```C++
// TODO(penpornk): Read the following configurations from a PluggableDevice
// callback instead of GPU environment variables: TF_GPU_THREAD_MODE,
// TF_GPU_THREAD_COUNT, TF_FORCE_GPU_ALLOC_GROWTH,
// TF_ENABLE_GPU_GARBAGE_COLLECTION, and TF_GPU_HOST_MEM_LIMIT_IN_MB.
```